libsupport/test/array-from-scalars.cpp

```diff
@@ -22,9 +22,28 @@ TestNulls() {
   return katana::ResultSuccess();
 }
 
+katana::Result<void>
+TestMixed() {
+  std::vector<std::shared_ptr<arrow::Scalar>> scalars(NUM_ENTRIES);
+  scalars[3] = std::make_shared<arrow::LargeStringScalar>("three");
+  scalars[4] = std::make_shared<arrow::LargeStringScalar>("four");
+  scalars[9] = std::make_shared<arrow::LargeStringScalar>("nine");
+  std::shared_ptr<arrow::Array> array =
+      KATANA_CHECKED(katana::ArrayFromScalars(scalars, arrow::large_utf8()));
+  // NB: length() is signed, size() is not - but they should both have small
+  // enough values that it doesn't matter
+  KATANA_LOG_VASSERT(
+      static_cast<size_t>(array->length()) == scalars.size(),
```

Comment:
It is better to ask to size_t or to the signed type?

Reply:
I'm not sure @danielmawhirter wrote that commit.

Reply:
We know by construction that both are within `[0, max<int64_t>]`, so I have no opinion on the casting convention

Reply:
Ok.  I don't ask to be a pain.  I just am a pain.

But I really do wonder about this because it comes up all the time with STL and Arrow.  I used to cast to `size_t`, but now I cast to `int64_t`.  I figure a negative number is more probable than a ridiculously large size.

Reply:
I would agree with that assessment. Casting in either direction does have some theoretical false positives thanks to the sign bit. I guess the most precise form would be something like:
```c++
bool equals(uint64_t u, int64_t i) {
  constexpr uint64_t high_bit = 1ULL << 63;
  if (u & high_bit) {
    // u is too high for an int64_t
    return false;
  } else if (i & high_bit) {
    // i is too low for a uint64_t
    return false;
  }
  // value is in range of both
  return static_cast<int64_t>(u) == i;
}

// code golf submission:
bool equals(uint64_t u, int64_t i) {
  return (!((u | i) & high_bit)) && static_cast<int64_t>(u) == i;
}
```

Or if we wait for c++20:
```c++
bool equals(uint64_t u, int64_t i) {
  return static_cast<std::make_signed_t<uint64_t>>(u) == i;
}
```

In light of those options, I consider some crazy edge case false positives to be an ok compromise. But I think you're right, the crazy edge case where `we have a negative and we shouldn't` is probably more common than `we have a really big value and we shouldn't`